src/main/kotlin/solve/rendering/engine/scene/Scene.kt

```diff
@@ -1,39 +1,58 @@
 package solve.rendering.engine.scene
 
-import solve.rendering.engine.core.renderers.Renderer
-
-class Scene(renderers: List<Renderer> = emptyList()) {
-    private val _renderObjects = mutableListOf<RenderObject>()
-    val renderObjects: List<RenderObject>
-        get() = _renderObjects
-
-    private val _renderers = renderers
-    val renderers: List<Renderer>
-        get() = _renderers
-
-    fun update(deltaTime: Float) {
-        _renderObjects.forEach { gameObject ->
-            if (gameObject.isDestroyed) {
-                removeRenderObject(gameObject)
-                return@forEach
-            }
-
-            gameObject.update(deltaTime)
-            _renderers.forEach { it.render() }
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import solve.rendering.engine.core.renderers.FramesRenderer
+import solve.rendering.engine.core.renderers.LandmarkLayerRenderer
+import solve.scene.model.Layer
+import java.util.concurrent.CopyOnWriteArrayList
+
+class Scene(val framesRenderer: FramesRenderer) {
+    private val _landmarkRenderers = CopyOnWriteArrayList<LandmarkLayerRenderer>()
+    private val _landmarkRenderersLayers = mutableMapOf<LandmarkLayerRenderer, Layer>()
+    val landmarkRenderers: List<LandmarkLayerRenderer>
+        get() = _landmarkRenderers
+    val landmarkLayerRendererLayers: Map<LandmarkLayerRenderer, Layer>
+        get() = _landmarkRenderersLayers
+
+    var isFramesRendererInitializing = false
+
+    fun update() {
+        render()
+    }
+
+    fun addLandmarkRenderer(renderer: LandmarkLayerRenderer, layer: Layer) {
+        _landmarkRenderers.add(renderer)
+        _landmarkRenderersLayers[renderer] = layer
+    }
+
+    fun clearLandmarkRenderers() {
+        landmarkRenderers.forEach { it.delete() }
+        _landmarkRenderers.clear()
+        _landmarkRenderersLayers.clear()
+    }
+
+    fun initializeFramesRenderer() {
+        CoroutineScope(Dispatchers.Default).launch {
```

Comment:
can it be canceled?

Reply:
There is a problem with rendering process. The `FramesRenderer` needs to initialize a textures array on a GPU side before it can render frames, and it requires some time. The problem is that if we try to execute `PlanesLayerRenderer` (which also loads textures to GPU) at that time, the textures array of `FramesRenderer` doesn't initialize correctly, and we get "black" frames. I spent a lot of time on this bug, and still haven't figured it out.